File: src/algorithms/heaps/operations/heap-replace-root/heap-replace-root.test.ts
import { describe, it, expect } from "vitest";
import { heapReplaceRoot } from "./sources/heap-replace-root.ts?fn";

/** Verify min-heap property: every parent ≤ both children. */
function isMinHeap(array: number[]): boolean {
  const size = array.length;
  for (let parentIdx = 0; parentIdx < Math.floor(size / 2); parentIdx++) {
    const leftIdx = 2 * parentIdx + 1;
    const rightIdx = 2 * parentIdx + 2;
    if (leftIdx < size && array[parentIdx]! > array[leftIdx]!) return false;
    if (rightIdx < size && array[parentIdx]! > array[rightIdx]!) return false;
  }
  return true;
}

describe("heapReplaceRoot", () => {
  it("returns the replaced value (original root)", () => {
    const result = heapReplaceRoot([1, 3, 5, 7, 9, 8, 6], 10) as {
      replacedValue: number;
      newHeap: number[];
    };
    expect(result.replacedValue).toBe(1);
  });

  it("produces a valid min-heap after replacing root with a large value", () => {
    const result = heapReplaceRoot([1, 3, 5, 7, 9, 8, 6], 10) as {
      replacedValue: number;
      newHeap: number[];
    };
    expect(isMinHeap(result.newHeap)).toBe(true);
  });

  it("new heap contains the new value", () => {
    const result = heapReplaceRoot([1, 3, 5, 7, 9, 8, 6], 10) as {
      replacedValue: number;
      newHeap: number[];
    };
    expect(result.newHeap.includes(10)).toBe(true);
    expect(result.newHeap.includes(1)).toBe(false);
  });

  it("preserves all elements (same multiset minus root, plus new value)", () => {
    const result = heapReplaceRoot([1, 3, 5, 7, 9, 8, 6], 10) as {
      replacedValue: number;
      newHeap: number[];
    };
    expect([...result.newHeap].sort((a, b) => a - b)).toEqual([3, 5, 6, 7, 8, 9, 10]);
  });

  it("replacing root with a value smaller than both children — no sift needed", () => {
    const result = heapReplaceRoot([1, 3, 5, 7, 9, 8, 6], 2) as {
      replacedValue: number;
      newHeap: number[];
    };
    expect(isMinHeap(result.newHeap)).toBe(true);
    expect(result.newHeap[0]).toBe(2);
    expect(result.replacedValue).toBe(1);
  });

  it("replacing root with the largest value sinks all the way to a leaf", () => {
    const result = heapReplaceRoot([1, 3, 5, 7, 9, 8, 6], 100) as {
      replacedValue: number;
      newHeap: number[];
    };
    expect(isMinHeap(result.newHeap)).toBe(true);
    expect(result.newHeap[0]).not.toBe(100);
  });

  it("handles a single-element heap", () => {
    const result = heapReplaceRoot([42], 7) as { replacedValue: number; newHeap: number[] };
    expect(result.replacedValue).toBe(42);
    expect(result.newHeap).toEqual([7]);
  });

  it("handles a two-element heap", () => {
    const result = heapReplaceRoot([1, 5], 10) as { replacedValue: number; newHeap: number[] };
    expect(result.replacedValue).toBe(1);
    expect(isMinHeap(result.newHeap)).toBe(true);
  });

  it("does not mutate the original array", () => {
    const input = [1, 3, 5, 7, 9, 8, 6];
    heapReplaceRoot(input, 10);
    expect(input).toEqual([1, 3, 5, 7, 9, 8, 6]);
  });
});

File: src/algorithms/heaps/operations/heap-replace-root/sources/heap-replace-root_test.py
import importlib

heap_replace_root = importlib.import_module("heap-replace-root").heap_replace_root


def is_min_heap(array):
    size = len(array)
    for parent_idx in range(size // 2):
        left_idx = 2 * parent_idx + 1
        right_idx = 2 * parent_idx + 2
        if left_idx < size and array[parent_idx] > array[left_idx]:
            return False
        if right_idx < size and array[parent_idx] > array[right_idx]:
            return False
    return True


def test_returns_replaced_value():
    result = heap_replace_root([1, 3, 5, 7, 9, 8, 6], 10)
    assert result["replaced_value"] == 1


def test_valid_min_heap_after_replace():
    result = heap_replace_root([1, 3, 5, 7, 9, 8, 6], 10)
    assert is_min_heap(result["new_heap"])


def test_new_value_in_heap():
    result = heap_replace_root([1, 3, 5, 7, 9, 8, 6], 10)
    assert 10 in result["new_heap"]
    assert 1 not in result["new_heap"]


def test_small_new_value_stays_at_root():
    result = heap_replace_root([1, 3, 5, 7, 9, 8, 6], 2)
    assert result["replaced_value"] == 1
    assert result["new_heap"][0] == 2


def test_large_value_sinks_to_leaf():
    result = heap_replace_root([1, 3, 5, 7, 9, 8, 6], 100)
    assert is_min_heap(result["new_heap"])
    assert result["new_heap"][0] != 100


def test_single_element():
    result = heap_replace_root([42], 7)
    assert result["replaced_value"] == 42
    assert result["new_heap"] == [7]


def test_two_element():
    result = heap_replace_root([1, 5], 10)
    assert result["replaced_value"] == 1
    assert is_min_heap(result["new_heap"])


if __name__ == "__main__":
    test_returns_replaced_value()
    test_valid_min_heap_after_replace()
    test_new_value_in_heap()
    test_small_new_value_stays_at_root()
    test_large_value_sinks_to_leaf()
    test_single_element()
    test_two_element()
    print("All tests passed!")
